README.md
# node-ember-cli-deploy-redis [![Build Status](https://travis-ci.org/blimmer/node-ember-cli-deploy-redis.svg?branch=master)](https://travis-ci.org/blimmer/node-ember-cli-deploy-redis)

ExpressJS middleware to fetch the current (or specified) revision of your Ember App deployed by [ember-cli-deploy](https://github.com/ember-cli/ember-cli-deploy).

## Why?

[ember-cli-deploy](https://github.com/ember-cli/ember-cli-deploy) is great. It allows you to run
multiple versions in production at the same time and view revisions without impacting users.
However, [the example provided](https://github.com/philipheinser/ember-lightning) uses [koa](http://koajs.com/)
and many of us are not. This package allows you to easily fetch current and specified `index.html`
revisions from [redis](http://redis.io) with [Express](expressjs.com) and other Node servers.

## Installation

It's important to choose the right version of this library to match the version of
ember-cli-deploy you're using.

| ember-cli-deploy version | node-ember-cli-deploy-redis |
|--------------------------|-----------------------------|
| pre 0.5                  | ^0.2.0 or lower             |
| 0.5 and beyond           | ^0.3.0 or newer             |

Make sure to look at the
[older documentation](https://github.com/blimmer/node-ember-cli-deploy-redis/blob/v0.2.0/README.md)
if you're on a pre 0.5 ember-cli-deploy release.
See [the changelog](https://github.com/blimmer/node-ember-cli-deploy-redis/blob/develop/CHANGELOG.md#030---2015-11-07)
for an upgrade guide.

## Usage

There are two main ways of using this library. For most simple Express servers, you'll
want to simply use the middleware. However, if you need more flexibility, you'll
want to use the internal `fetch` methods, with custom logic.

### ExpressJS Middleware

1. `require` the package
2. `use` the package in your app

#### Example

```javascript
const express = require('express');
const app = express();

const nodeEmberCliDeployRedis = require('node-ember-cli-deploy-redis');
app.use('/*', nodeEmberCliDeployRedis('myapp:index', {
  host: 'redis.example.org',    // default is localhost
  port: 6929,                   // default is 6379
  password: 'passw0rd!',        // default is undefined
  db: 0                         // default is undefined
}));
```

<hr>

### Custom Fetch Method

1. `require` the package
2. Use the `fetchIndex` method
3. Render the index string as you wish.

#### Example

```javascript
const express = require('express');
const app = express();

const fetchIndex = require('node-ember-cli-deploy-redis/fetch');

app.get('/', function(req, res) {
    fetchIndex(req, 'myapp:index', {
      host: 'redis.example.org',
      port: 6929,
      password: 'passw0rd!',
      db: 0
    }).then(function (indexHtml) {
    indexHtml = serverVarInjectHelper.injectServerVariables(indexHtml, req);
    res.status(200).send(indexHtml);
  }).catch(function(err) {
    res.status(500).send('Oh noes!\n' + err.message);
  });
});
```
Check out [location-aware-ember-server](https://github.com/blimmer/location-aware-ember-server) for a running example.

## Documentation

### `nodeEmberCliDeployRedis(keyPrefix, connectionInfo, options)` (middleware constructor)

* keyPrefix (required) - the application name, specified for ember deploy
   the keys in redis are prefaced with this name. For instance, if your redis keys are `my-app:index:current`, you'd pass `my-app:index`.
* connectionInfo (required) - the configuration to connect to redis.
   internally, this library uses [ioredis](https://github.com/luin/ioredis), so pass a configuration supported by ioredis. please see their README for more information.
* options (optional) - a hash of params to override [the defaults](https://github.com/blimmer/node-ember-cli-deploy-redis/blob/develop/README.md#options)

### `fetchIndex(request, keyPrefix, connectionInfo, options)`

Arguments

* request (required) - the request object
   the request object is used to check for the presence of `revisionQueryParam`
* keyPrefix (required) - the application name, specified for ember deploy
   the keys in redis are prefaced with this name. For instance, if your redis keys are `my-app:index:current`, you'd pass `my-app:index`.
* connectionInfo (required) - the configuration to connect to redis.
   internally, this library uses [ioredis](https://github.com/luin/ioredis), so pass a configuration supported by ioredis.
* options (optional) - a hash of params to override [the defaults](https://github.com/blimmer/node-ember-cli-deploy-redis/blob/develop/README.md#options)

Returns

* a `Promise`
   when resolved, it returns the requested `index.html` string
   when failed, it returns an [EmberCliDeployError](https://github.com/blimmer/node-ember-cli-deploy-redis/blob/develop/errors/ember-cli-deploy-error.js).

### options

* `revisionQueryParam` (defaults to `index_key`)
   the query parameter to specify a revision (e.g. `http://example.org/?index_key=abc123`). the key will be automatically prefaced with your `keyPrefix` for security.
* `memoize` (defaults to `false`)
   enable memoizing Redis `get`s. see [the memoization section](#Memoization) for more details.
* `memoizeOpts` ([see defaults](https://github.com/blimmer/node-ember-cli-deploy-redis/blob/master/fetch.js#L18))
   customize memoization parameters. see [the memoization section](#Memoization) for more details.

## Memoization

Since the majority of the requests will be serving the `current` version of your
app, you can enable memoization to reduce the load on Redis. By default, memoization
is disabled. To enable it, simply pass:

```javascript
memoize: true
```

in your options hash. Additionally, you can pass options to the underlying memoization
library ([memoizee](https://github.com/medikoo/memoizee)). Check out their documentation,
and the [defaults](https://github.com/blimmer/node-ember-cli-deploy-redis/blob/master/fetch.js#L18)
for this library.

### Example

```javascript
app.use('/*', nodeEmberCliDeployRedis(
  'myapp:index',
  { host: 'redis.example.org' },
  { memoize: true },
));
```

## Testing

In order to facilitate unit testing and/or integration testing this
library exports a mockable redis api.  You will need to use a
dependency injection framework such as
[rewire](https://github.com/jhnns/rewire) to activate this testing api.

### Usage with rewire (mocha syntax)

```javascript
// my-module.js
const fetchIndex = require('node-ember-cli-deploy-redis/fetch');
const indexWrapper = function(req, res) {
  return fetchIndex(req, 'app', {
    // real redis config
  }).then(function (indexHtml)) {
    // do something with index
  });
};
module.exports = indexWrapper;

// my-module-test.js
const redisTestApi = require('node-ember-cli-deploy-redis/test/helpers/test-api');
const fetchIndex = rewire('node-ember-cli-deploy-redis/fetch');
const redis = redisTestApi.ioRedisApi;
const myModule = rewire('my-module');

describe('my module', function() {
  afterEach(function() {
    fetchIndex.__set__('_initialized', false);
  });

  it('grabs my content', function() {
    // inject mocked content
    myModule.__set__('fetchIndex', fetchIndex);
    fetchIndex.__set__('ioRedis', redis);
    redis.set('app:abc123', "<html><body><h1>hello test world</h1></body></html>");
    myModule(req, res).then(function(){
      // assertions here
    })
  });
});
```

## Notes

* Don't create any other redis keys you don't want exposed to the public under your `keyPrefix`.

## Contributing

Comments/PRs/Issues are welcome!

### Cloning

This project utilizes git submodules. Please use the following command to clone:

```console
git clone --recurse-submodules https://github.com/blimmer/node-ember-cli-deploy-redis.git
```

### Running Project Unit Tests

```console
npm test
```

### Running Project Smoke Tests

```console
npm run smoke-test
```


CHANGELOG.md
# Change Log

## [1.0.1](https://github.com/blimmer/node-ember-cli-deploy-redis/tree/1.0.1) (2019-08-31)
[Full Changelog](https://github.com/blimmer/node-ember-cli-deploy-redis/compare/v1.0.0...1.0.1)

**Merged pull requests:**

- Bump lodash from 4.17.11 to 4.17.13 [\#24](https://github.com/blimmer/node-ember-cli-deploy-redis/pull/24) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump eslint-utils from 1.3.1 to 1.4.2 [\#23](https://github.com/blimmer/node-ember-cli-deploy-redis/pull/23) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump js-yaml from 3.12.0 to 3.13.1 [\#22](https://github.com/blimmer/node-ember-cli-deploy-redis/pull/22) ([dependabot[bot]](https://github.com/apps/dependabot))

## [v1.0.0](https://github.com/blimmer/node-ember-cli-deploy-redis/tree/v1.0.0) (2018-11-28)
[Full Changelog](https://github.com/blimmer/node-ember-cli-deploy-redis/compare/v0.4.1...v1.0.0)

**Merged pull requests:**

- \[POTENITALLY BREAKING CHANGES\] Upgrade memoizee and ioredis [\#21](https://github.com/blimmer/node-ember-cli-deploy-redis/pull/21) ([blimmer](https://github.com/blimmer))
- Use native Promise instead of Bluebird. [\#20](https://github.com/blimmer/node-ember-cli-deploy-redis/pull/20) ([blimmer](https://github.com/blimmer))
- \[BREAKING CHANGE\] Pass error back to app from middleware instead of rendering a `500` [\#19](https://github.com/blimmer/node-ember-cli-deploy-redis/pull/19) ([blimmer](https://github.com/blimmer))
- Convert smoke-test suite to bats [\#18](https://github.com/blimmer/node-ember-cli-deploy-redis/pull/18) ([blimmer](https://github.com/blimmer))
- Add smoke tests [\#17](https://github.com/blimmer/node-ember-cli-deploy-redis/pull/17) ([blimmer](https://github.com/blimmer))
- Upgrade dev dependencies, supported node \(\>=6\) and replace jshint [\#16](https://github.com/blimmer/node-ember-cli-deploy-redis/pull/16) ([blimmer](https://github.com/blimmer))
- Remove `database` deprecation and fallback. [\#15](https://github.com/blimmer/node-ember-cli-deploy-redis/pull/15) ([blimmer](https://github.com/blimmer))

## [v0.4.1](https://github.com/blimmer/node-ember-cli-deploy-redis/tree/v0.4.1) (2016-04-03)
[Full Changelog](https://github.com/blimmer/node-ember-cli-deploy-redis/compare/0.4.0...v0.4.1)

**Merged pull requests:**

- Dependency upgrades. [\#13](https://github.com/blimmer/node-ember-cli-deploy-redis/pull/13) ([blimmer](https://github.com/blimmer))

## [0.4.0](https://github.com/blimmer/node-ember-cli-deploy-redis/tree/0.4.0) (2016-02-05)
[Full Changelog](https://github.com/blimmer/node-ember-cli-deploy-redis/compare/v0.3.0...0.4.0)

**Implemented enhancements:**

- Throttle calls to redis.get [\#4](https://github.com/blimmer/node-ember-cli-deploy-redis/issues/4)

**Merged pull requests:**

- IORedis / Memoization Support [\#12](https://github.com/blimmer/node-ember-cli-deploy-redis/pull/12) ([blimmer](https://github.com/blimmer))

## [v0.3.0](https://github.com/blimmer/node-ember-cli-deploy-redis/tree/v0.3.0) (2015-11-08)
[Full Changelog](https://github.com/blimmer/node-ember-cli-deploy-redis/compare/v0.2.0...v0.3.0)

**Merged pull requests:**

- ember cli deploy 0.5.0 compatibility [\#10](https://github.com/blimmer/node-ember-cli-deploy-redis/pull/10) ([blimmer](https://github.com/blimmer))

## [v0.2.0](https://github.com/blimmer/node-ember-cli-deploy-redis/tree/v0.2.0) (2015-07-08)
[Full Changelog](https://github.com/blimmer/node-ember-cli-deploy-redis/compare/v0.1.1...v0.2.0)

**Closed issues:**

- api inconsistency [\#6](https://github.com/blimmer/node-ember-cli-deploy-redis/issues/6)

**Merged pull requests:**

- v0.2.0 RC [\#8](https://github.com/blimmer/node-ember-cli-deploy-redis/pull/8) ([blimmer](https://github.com/blimmer))
- Fix API inconsistency between fetch and middleware [\#7](https://github.com/blimmer/node-ember-cli-deploy-redis/pull/7) ([knownasilya](https://github.com/knownasilya))

## [v0.1.1](https://github.com/blimmer/node-ember-cli-deploy-redis/tree/v0.1.1) (2015-06-02)
[Full Changelog](https://github.com/blimmer/node-ember-cli-deploy-redis/compare/v0.1.0...v0.1.1)

**Merged pull requests:**

- Expose the test api for mocking [\#5](https://github.com/blimmer/node-ember-cli-deploy-redis/pull/5) ([ronco](https://github.com/ronco))

## [v0.1.0](https://github.com/blimmer/node-ember-cli-deploy-redis/tree/v0.1.0) (2015-06-01)
[Full Changelog](https://github.com/blimmer/node-ember-cli-deploy-redis/compare/v0.0.1...v0.1.0)

**Implemented enhancements:**

- Allow using as a middleware [\#2](https://github.com/blimmer/node-ember-cli-deploy-redis/issues/2)
- Allow working with then-redis and node\_redis [\#1](https://github.com/blimmer/node-ember-cli-deploy-redis/issues/1)

**Merged pull requests:**

- Rework to have internal then-redis dependency. [\#3](https://github.com/blimmer/node-ember-cli-deploy-redis/pull/3) ([blimmer](https://github.com/blimmer))

## [v0.0.1](https://github.com/blimmer/node-ember-cli-deploy-redis/tree/v0.0.1) (2015-05-24)


\* *This Change Log was automatically generated by [github_changelog_generator](https://github.com/skywinder/Github-Changelog-Generator)*

test/fetch-test.js
const { expect } = require('chai');
require('chai').use(require("sinon-chai"));

const { describe, it, before, after, beforeEach, afterEach } = require('mocha');
const context = describe;

const sinon =  require('sinon');
const rewire = require('rewire');

const fetchIndex = rewire('../fetch');

const basicReq = {
  query: {}
};

const testApi = require('./helpers/test-api');
const redisClientApi = testApi.ioRedisClientApi;
const ioRedisApi = testApi.ioRedisApi;

describe('fetch', function() {
  let sandbox;
  before(function() {
    sandbox = sinon.createSandbox();
    fetchIndex.__set__('ioRedis', ioRedisApi);
  });

  afterEach(function() {
    fetchIndex.__set__('initialized', false);
    sandbox.restore();
  });

  describe('_initialize', function () {
    let _initialize;
    before(function() {
      _initialize = fetchIndex.__get__('_initialize');
    });
    after(function() {
      fetchIndex.__set__('ioRedis', ioRedisApi);
    });

    describe('redis client initialize', function() {
      let ioRedisInitStub;
      beforeEach(function() {
        ioRedisInitStub = sandbox.stub();
        fetchIndex.__set__('ioRedis', ioRedisInitStub);
      });

      it('sets up redis (defaults)', function() {
        _initialize();

        expect(ioRedisInitStub).to.have.been.calledOnce;
        expect(ioRedisInitStub).to.have.been.calledWithNew;
        expect(ioRedisInitStub).to.have.been.calledOnceWith({
          host: "127.0.0.1",
          port: 6379
        });
      });

      it('sets up redis (config passed)', function() {
        let configString = 'redis://h:passw0rd@example.org:6929';
        _initialize(configString);

        expect(ioRedisInitStub).to.have.been.calledOnce;
        expect(ioRedisInitStub).to.have.been.calledWithNew;
        expect(ioRedisInitStub).to.have.been.calledOnceWith(configString);
      });
    });

    describe('memoization', function() {
      let memoizeStub;
      beforeEach(function() {
        memoizeStub = sandbox.stub();
        fetchIndex.__set__('memoize', memoizeStub);
      });

      after(function() {
        fetchIndex.__set__('memoize', require('memoizee'));
      });

      context('not enabled (default)', function() {
        it('does not enable memoize redis.get', function() {
          _initialize();
          expect(memoizeStub.called).to.be.false;
        });
      });

      context('enabled', function() {
        it('passes default options to memoize', function() {
          _initialize({}, { memoize: true });

          expect(memoizeStub).to.have.been.calledOnceWith(undefined, {
            maxAge: 5000,
            preFetch: true,
            max: 4,
            async: false,
            length: 1,
          });
        });

        it('allows overriding existing properties', function() {
          let myOpts = {
            maxAge: 10000,
            preFetch: 0.6,
            max: 2,
          };

          _initialize({}, {
            memoize: true,
            memoizeOpts: myOpts,
          });

          expect(memoizeStub).calledOnceWith(undefined, {
            maxAge: 10000,
            preFetch: 0.6,
            max: 2,
            async: false,
            length: 1,
          });
        });

        it('allows adding additional memoizee options', function() {
          const myDispose = function() {
            // some custom dispose logic because I'm a masochist
          };
          let myOpts = {
            dispose: myDispose
          };

          _initialize({}, {
            memoize: true,
            memoizeOpts: myOpts,
          });

          expect(memoizeStub).calledOnceWith(undefined, {
            async: false,
            dispose: myDispose,
            length: 1,
            max: 4,
            maxAge: 5000,
            preFetch: true
          });
        });

        it('does not allow overriding async flag', function() {
          let myOpts = {
            async: true,
          };

          _initialize({}, {
            memoize: true,
            memoizeOpts: myOpts,
          });

          expect(memoizeStub).to.have.been.calledOnceWith(undefined, {
            maxAge: 5000,
            preFetch: true,
            max: 4,
            async: false,
            length: 1,
          });
        });

        it('does not allow overriding the length property', function() {
          let myOpts = {
            length: 2,
          };

          _initialize({}, {
            memoize: true,
            memoizeOpts: myOpts,
          });

          expect(memoizeStub).to.have.been.calledOnceWith(undefined, {
            maxAge: 5000,
            preFetch: true,
            max: 4,
            async: false,
            length: 1,
          });
        });
      });
    });

    describe('revisionQueryParam', function() {
      it('has a default revisionQueryParam', function() {
        _initialize();

        let opts = fetchIndex.__get__('opts');
        expect(opts.revisionQueryParam).to.equal('index_key');
      });

      it('allows override of revisionQueryParam', function() {
        _initialize({}, {revisionQueryParam: 'foobar'});

        let opts = fetchIndex.__get__('opts');
        expect(opts.revisionQueryParam).to.equal('foobar');
      });
    });

    it('sets initialized flag', function() {
      expect(fetchIndex.__get__('initialized')).to.be.false;
      _initialize();
      expect(fetchIndex.__get__('initialized')).to.be.true;
    });
  });

  describe('fetchIndex', function() {
    let redis, redisSpy;

    before(function() {
      redis = redisClientApi;
    });

    beforeEach(function() {
      redisSpy = sandbox.spy(redis, 'get');
    });

    afterEach(function() {
      redis.flushall();
    });

    it('normalizes spaces in revisionQueryParam', function(done) {
      const req = {
        query: {
          index_key: 'abc 123'
        }
      };

      redis.set('myapp:index:abc123', 'foo').then(function(){
        fetchIndex(req, 'myapp:index').then(function() {
          expect(redisSpy).to.have.been.calledWith('myapp:index:abc123');
          expect(redisSpy).to.not.have.been.calledWith('myapp:index:abc 123');
          done();
        }).catch(function(err) {
          done(err);
        });
      });

    });

    it('removes special chars revisionQueryParam', function(done) {
      const req = {
        query: {
          index_key: 'ab@*#!c(@)123'
        }
      };

      redis.set('myapp:index:abc123', 'foo').then(function(){
        fetchIndex(req, 'myapp:index').then(function() {
          expect(redisSpy).to.have.been.calledWith('myapp:index:abc123');
          expect(redisSpy).to.not.have.been.calledWith('myapp:index:ab@*#!c(@)123');
          done();
        }).catch(function(err) {
          done(err);
        });
      });
    });

    it('fails the promise with a critical error if keyPrefix:current is not present', function(done) {
      redis.del('myapp:index:current').then(function(){
        fetchIndex(basicReq, 'myapp:index').then(function() {
          done("Promise should not have resolved.");
        }).catch(function(err) {
          expect(redisSpy).to.have.been.calledWith('myapp:index:current');
          expect(err.critical).to.be.true;
          done();
        });
      });
    });

    it('fails the promise with a critical error if revision pointed to by keyPrefix:current is not present', function(done) {
      redis.set('myapp:index:current', 'abc123').then(function(){
        return redis.del('myapp:index:abc123');
      }).then(function(){
        fetchIndex(basicReq, 'myapp:index').then(function() {
          done("Promise should not have resolved.");
        }).catch(function(err) {
          expect(redisSpy).to.have.been.calledWith('myapp:index:current');
          expect(redisSpy).to.have.been.calledWith('myapp:index:abc123');
          expect(err.critical).to.be.true;
          done();
        });
      });
    });

    it('fails the promise with a non-critical error if revision requestd by query param is not present', function(done) {
      const req = {
        query: {
          index_key: 'abc123'
        }
      };
      redis.del('myapp:index:abc123').then(function(){
        fetchIndex(req, 'myapp:index').then(function() {
          done("Promise should not have resolved.");
        }).catch(function(err) {
          expect(redisSpy).to.have.been.calledWith('myapp:index:abc123');
          expect(err.critical).to.be.false;
          done();
        });
      });
    });

    it('resolves the promise with the index html requested', function(done) {
      const currentHtmlString = '<html><body>1</body></html>';
      Promise.all([
        redis.set('myapp:index:current', 'abc123'),
        redis.set('myapp:index:abc123', currentHtmlString),
      ]).then(function(){
        fetchIndex(basicReq, 'myapp:index').then(function(html) {
          expect(redisSpy).to.have.been.calledWith('myapp:index:current');
          expect(redisSpy).to.have.been.calledWith('myapp:index:abc123');
          expect(html).to.equal(currentHtmlString);
          done();
        }).catch(function() {
          done("Promise should not have failed.");
        });
      });
    });

    it('resolves the promise with the index html requested (specific revision)', function(done) {
      const currentHtmlString = '<html><body>1</body></html>';
      const newDeployHtmlString = '<html><body>2</body></html>';
      const req = {
        query: {
          index_key: 'def456'
        }
      };
      Promise.all([
        redis.set('myapp:index:current', 'abc123'),
        redis.set('myapp:index:abc123', currentHtmlString),
        redis.set('myapp:index:def456', newDeployHtmlString)
      ]).then(function(){
        fetchIndex(req, 'myapp:index').then(function(html) {
          expect(redisSpy).to.not.have.been.calledWith('myapp:index:current');
          expect(redisSpy).to.have.been.calledWith('myapp:index:def456');
          expect(html).to.equal(newDeployHtmlString);
          done();
        }).catch(function() {
          done("Promise should not have failed.");
        });
      });
    });

    it('memoizes results from redis when turned on', function(done) {
      const currentHtmlString = '<html><body>1</body></html>';
      Promise.all([
        redis.set('myapp:index:current', 'abc123'),
        redis.set('myapp:index:abc123', currentHtmlString),
        fetchIndex(basicReq, 'myapp:index', null, { memoize: true }),
        fetchIndex(basicReq, 'myapp:index', null, { memoize: true }),
        fetchIndex(basicReq, 'myapp:index', null, { memoize: true }),
      ]).then(function(){
        expect(redisSpy).to.have.been.calledWith('myapp:index:current');
        expect(redisSpy).to.have.been.calledWith('myapp:index:abc123');
        done();
      });
    });
  });
});


test/index-test.js
var expect    = require('chai').expect;
var { describe, before, beforeEach, afterEach, it } = require('mocha');

var sinon     = require('sinon');
var httpMocks = require('node-mocks-http');
var rewire    = require('rewire');

var middleware = rewire('../index');
var EmberCliDeployError = require('../errors/ember-cli-deploy-error');

var htmlString = '<html><body>1</body></html>';

describe('express middleware', function() {
  var sandbox, req, res, fetchIndexStub;
  before(function() {
    sandbox = sinon.createSandbox();
  });

  beforeEach(function() {
    req = httpMocks.createRequest();
    res = httpMocks.createResponse();

    fetchIndexStub = sandbox.stub();
    middleware.__set__('fetchIndex', fetchIndexStub);
  });

  afterEach(function() {
    sandbox.restore();
  });

  describe('success', function() {
    it('returns a 200 and sends the html', function(done) {
      fetchIndexStub.returns(Promise.resolve(htmlString));

      middleware()(req, res).then(function() {
        expect(res.statusCode).to.equal(200);
        var data = res._getData();
        expect(data).to.equal(htmlString);
        done();
      });
    });
  });

  describe('failure', function() {
    it('calls the `next` function with the error', function(done) {
      var error = new EmberCliDeployError();
      fetchIndexStub.returns(Promise.reject(error));

      function nextExpectation() {
        expect(arguments).to.have.length(1);
        expect(arguments[0]).to.equal(error);
        done();
      }

      middleware()(req, res, nextExpectation).then(function() {
        done('Promise should not have resolved');
      });
    });
  });
});


test/helpers/test-api.js
const ioRedisClientApi = {
  _storage: {},
  get: function(key){
    return Promise.resolve(this._storage[key]);
  },
  set: function(key, value){
    this._storage[key] = value;
    return Promise.resolve(value);
  },
  del: function(key){
    delete this._storage[key];
    return Promise.resolve();
  },
  flushall: function(){
    this._storage = {};
    return Promise.resolve();
  }
};

const ioRedisApi = function() {
  return ioRedisClientApi;
};

module.exports = {
  ioRedisClientApi: ioRedisClientApi,
  ioRedisApi: ioRedisApi
};


smoke-test/run.sh
#! /bin/bash

set -e

# This script starts up a redis database and seeds it with some fixture data
# to do a quick smoke-test that things are working as expected.

SCRIPT_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"

REDIS_DOCKER_IMAGE='redis:5-alpine'
DOCKER_CONTAINER_NAME='node-ember-cli-deploy-redis-smoke-test'
EXPRESS_APP_PID=''

_start-redis() {
  echo "Starting redis container..."
  docker run -p 6379:6379 --name $DOCKER_CONTAINER_NAME -d $REDIS_DOCKER_IMAGE
}

_start-express-app() {
  echo "Starting express app..."
  pushd "$SCRIPT_DIR/express"
  npm install --no-package-lock
  NODE_ENV=production npm start &
  EXPRESS_APP_PID=$!

  echo "Waiting for express app to start up..."
  while ! nc -z localhost 3000; do
    sleep 0.1
  done
  echo "Express app is started!"

  popd
}

# If this gets more complex, we might want to use BATS
# https://github.com/sstephenson/bats
_test() {
  echo "Running tests..."
  DOCKER_CONTAINER_NAME=$DOCKER_CONTAINER_NAME "$SCRIPT_DIR"/test-helper/bats/bin/bats "$SCRIPT_DIR"/test.bats
}

_stop-redis() {
  echo "Killing redis container..."
  docker stop $DOCKER_CONTAINER_NAME > /dev/null
  docker rm $DOCKER_CONTAINER_NAME > /dev/null
}

_stop-express-app() {
  if [ $EXPRESS_APP_PID != "" ]; then
   echo "Killing express app..."
   kill $EXPRESS_APP_PID
  fi
}

_cleanup() {
  _stop-express-app
  _stop-redis
}

main() {
  _start-redis
  _start-express-app
  _test
}

trap _cleanup EXIT
main


fetch.js
const _defaultsDeep = require('lodash/defaultsDeep');

const EmberCliDeployError = require('./errors/ember-cli-deploy-error');

const ioRedis = require('ioredis');
const memoize = require('memoizee');
let redisClient;

const defaultConnectionInfo = {
  host: "127.0.0.1",
  port: 6379
};

const _defaultOpts = {
  revisionQueryParam: 'index_key',
  memoize: false,
  memoizeOpts: {
    maxAge:   5000, // ms
    preFetch: true,
    max:      4,    // a sane default (current pointer, current html and two indexkeys in cache)
  }
};

let opts;
function _getOpts(opts) {
  opts = opts || {};
  return _defaultsDeep({}, opts, _defaultOpts);
}

let initialized = false;
function _initialize(connectionInfo, passedOpts) {
  opts = _getOpts(passedOpts);
  let config = connectionInfo ? connectionInfo : defaultConnectionInfo;

  redisClient = new ioRedis(config);

  if (opts.memoize === true) {
    let memoizeOpts = opts.memoizeOpts;
    memoizeOpts.async = false; // this should never be overwritten by the consumer
    memoizeOpts.length = 1;

    redisClient.get = memoize(redisClient.get, memoizeOpts);
  }

  initialized = true;
}

function fetchIndex(req, keyPrefix, connectionInfo, passedOpts) {
  if (!initialized) {
    _initialize(connectionInfo, passedOpts);
  }

  let indexkey;
  if (req.query[opts.revisionQueryParam]) {
    let queryKey = req.query[opts.revisionQueryParam].replace(/[^A-Za-z0-9]/g, '');
    indexkey = `${keyPrefix}:${queryKey}`;
  }

  let customIndexKeyWasSpecified = !!indexkey;
  function retrieveIndexKey(){
    if (indexkey) {
      return Promise.resolve(indexkey);
    } else {
      return redisClient.get(keyPrefix + ":current").then(function(result){
        if (!result) { throw new Error(); }
        return keyPrefix + ":" + result;
      }).catch(function(){
        throw new EmberCliDeployError("There's no " + keyPrefix + ":current revision. The site is down.", true);
      });
    }
  }

  return retrieveIndexKey().then(function(indexkey){
    return redisClient.get(indexkey);
  }).then(function(indexHtml){
      if (!indexHtml) { throw new Error(); }
      return indexHtml;
  }).catch(function(err){
    if (err.name === 'EmberCliDeployError') {
      throw err;
    } else {
      throw new EmberCliDeployError("There's no " + indexkey + " revision. The site is down.", !customIndexKeyWasSpecified);
    }
  });
}

module.exports = fetchIndex;
